src/lib/ffi/ffi_tpm2.cpp

```diff
@@ -0,0 +1,205 @@
+   return Botan::TPM2::SessionBundle((s1 != nullptr) ? safe_get(s1).session : nullptr,
+                                     (s2 != nullptr) ? safe_get(s2).session : nullptr,
+                                     (s3 != nullptr) ? safe_get(s3).session : nullptr);
+}
+
+}  // namespace
+
+extern "C" {
+
+#endif
+
+int botan_tpm2_supports_crypto_backend() {
+#if defined(BOTAN_HAS_TPM2)
+   return Botan::TPM2::Context::supports_botan_crypto_backend() ? 1 : 0;
+#else
+   return 0;
+#endif
+}
+
+int botan_tpm2_ctx_init(botan_tpm2_ctx_t* ctx_out, const char* tcti_nameconf) {
+#if defined(BOTAN_HAS_TPM2)
+   return ffi_guard_thunk(__func__, [=]() -> int {
+      if(ctx_out == nullptr) {
+         return BOTAN_FFI_ERROR_NULL_POINTER;
+      }
+      auto ctx = std::make_unique<botan_tpm2_ctx_wrapper>();
+
+      auto tcti = [=]() -> std::optional<std::string> {
+         if(tcti_nameconf == nullptr) {
+            return {};
+         } else {
+            return std::string(tcti_nameconf);
+         }
+      }();
+
+      ctx->ctx = Botan::TPM2::Context::create(std::move(tcti));
+      *ctx_out = new botan_tpm2_ctx_struct(std::move(ctx));
+      return BOTAN_FFI_SUCCESS;
+   });
+#else
+   BOTAN_UNUSED(ctx_out, tcti_nameconf);
+   return BOTAN_FFI_ERROR_NOT_IMPLEMENTED;
+#endif
+}
+
+int botan_tpm2_ctx_init_ex(botan_tpm2_ctx_t* ctx_out, const char* tcti_name, const char* tcti_conf) {
+#if defined(BOTAN_HAS_TPM2)
+   return ffi_guard_thunk(__func__, [=]() -> int {
+      if(ctx_out == nullptr) {
+         return BOTAN_FFI_ERROR_NULL_POINTER;
+      }
+      auto ctx = std::make_unique<botan_tpm2_ctx_wrapper>();
+
+      auto tcti_name_str = [=]() -> std::optional<std::string> {
+         if(tcti_name == nullptr) {
+            return {};
+         } else {
+            return std::string(tcti_name);
+         }
+      }();
+
+      auto tcti_conf_str = [=]() -> std::optional<std::string> {
+         if(tcti_conf == nullptr) {
+            return {};
+         } else {
+            return std::string(tcti_conf);
+         }
+      }();
+
+      ctx->ctx = Botan::TPM2::Context::create(std::move(tcti_name_str), std::move(tcti_conf_str));
+      *ctx_out = new botan_tpm2_ctx_struct(std::move(ctx));
+      return BOTAN_FFI_SUCCESS;
+   });
+#else
+   BOTAN_UNUSED(ctx_out, tcti_name, tcti_conf);
+   return BOTAN_FFI_ERROR_NOT_IMPLEMENTED;
+#endif
+}
+
+int botan_tpm2_ctx_enable_crypto_backend(botan_tpm2_ctx_t ctx, botan_rng_t rng) {
+#if defined(BOTAN_HAS_TPM2)
+   return BOTAN_FFI_VISIT(ctx, [=](botan_tpm2_ctx_wrapper& ctx_wrapper) -> int {
+      Botan::RandomNumberGenerator& rng_ref = safe_get(rng);
+
+      // The lifetime of the RNG used for the crypto backend should be managed
+      // by the TPM2::Context. Here, we just need to trust the user that they
+      // keep the passed-in RNG instance intact for the lifetime of the context.
+      std::shared_ptr<Botan::RandomNumberGenerator> rng_ptr(&rng_ref, [](auto*) {});
```

Comment:
:/ We could have this function steal (move) the RNG but that’s not otherwise a convention used in the FFI.

Another alternative would be to remove the choice of which RNG is used, and just wrap eg system_rng. However that goes against the library-wide convention that the application should in all contexts be able to specify what RNG they prefer to use.

Reply:
I mean, this solution works and is acceptable in my opinion. Certainly better than hardcoding the RNG.

Reply:
No objections to doing it this way, just wanted to offer any alternative approaches I could think of.